File: src/main/go/leetcode/editor/cn/31/NextPermutation.go
//31.next-permutation
//实现获取 下一个排列 的函数，算法需要将给定数字序列重新排列成字典序中下一个更大的排列。
//
// 如果不存在下一个更大的排列，则将数字重新排列成最小的排列（即升序排列）。
//
// 必须 原地 修改，只允许使用额外常数空间。
//
//
//
// 示例 1：
//
//
//输入：nums = [1,2,3]
//输出：[1,3,2]
//
//
// 示例 2：
//
//
//输入：nums = [3,2,1]
//输出：[1,2,3]
//
//
// 示例 3：
//
//
//输入：nums = [1,1,5]
//输出：[1,5,1]
//
//
// 示例 4：
//
//
//输入：nums = [1]
//输出：[1]
//
//
//
//
// 提示：
//
//
// 1 <= nums.length <= 100
// 0 <= nums[i] <= 100
//
// Related Topics 数组
// 👍 1173 👎 0

package main

import (
	"fmt"
	"sort"
)

//leetcode submit region begin(Prohibit modification and deletion)

func nextPermutation(nums []int) {
	var l = len(nums)
	var r = 0
	//从找到最后一个升序点
	for e := l - 1; e > 0; e-- {
		if nums[e] > nums[e-1] {
			r = e
			break
		}
	}
	if r > 0 {
		//从升序点(num[i]<num[j])将数组分割两部分，i属于左半部分，j属于右半部分
		//我们的思路是，左部分变为大于当前值的最小值，右部分变为最小值，合并后就是大于当前值的最小值
		//对于左半部分，将i位置替换为右半部分大于num[i]的最小值
		//对于右半部分，直接升序排列
		var t = r
		for i := r; i < l; i++ {
			//大于nums[r-1]的最小值
			if nums[i] > nums[r-1] && nums[i] < nums[t] {
				t = i
			}
		}
		minSwap(nums, r-1, t)
	} else { //如果不存在升序，则已经是最大值
		sort.Ints(nums)
	}
}

/*
*
begin和end交换位置，然后begin之后升序排序
*/
func minSwap(nums []int, begin int, end int) {
	//交换数据
	nums[begin], nums[end] = nums[end], nums[begin]
	//将begin之后的数据排序
	var arr = nums[begin+1:]
	sort.Ints(arr)
	for i, j := range arr {
		nums[begin+1+i] = j
	}
}

//leetcode submit region end(Prohibit modification and deletion)

// test from here
func main() {
	//var a = []int{4, 2, 0, 2, 3, 2, 0}
	var a = []int{1, 3, 2}
	nextPermutation(a)
	fmt.Println(a)
}

File: src/main/java/leetcode/editor/en/NextPermutation.java
//31.next-permutation
/**
 * A permutation of an array of integers is an arrangement of its members into a
 * sequence or linear order.
 * <p>
 * <p>
 * For example, for arr = [1,2,3], the following are all the permutations of arr:
 * [1,2,3], [1,3,2], [2, 1, 3], [2, 3, 1], [3,1,2], [3,2,1].
 * <p>
 * <p>
 * The next permutation of an array of integers is the next lexicographically
 * greater permutation of its integer. More formally, if all the permutations of the
 * array are sorted in one container according to their lexicographical order, then
 * the next permutation of that array is the permutation that follows it in the
 * sorted container. If such arrangement is not possible, the array must be rearranged
 * as the lowest possible order (i.e., sorted in ascending order).
 * <p>
 * <p>
 * For example, the next permutation of arr = [1,2,3] is [1,3,2].
 * Similarly, the next permutation of arr = [2,3,1] is [3,1,2].
 * While the next permutation of arr = [3,2,1] is [1,2,3] because [3,2,1] does
 * not have a lexicographical larger rearrangement.
 * <p>
 * <p>
 * Given an array of integers nums, find the next permutation of nums.
 * <p>
 * The replacement must be in place and use only constant extra memory.
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * Input: nums = [1,2,3]
 * Output: [1,3,2]
 * <p>
 * <p>
 * Example 2:
 * <p>
 * <p>
 * Input: nums = [3,2,1]
 * Output: [1,2,3]
 * <p>
 * <p>
 * Example 3:
 * <p>
 * <p>
 * Input: nums = [1,1,5]
 * Output: [1,5,1]
 * <p>
 * <p>
 * <p>
 * Constraints:
 * <p>
 * <p>
 * 1 <= nums.length <= 100
 * 0 <= nums[i] <= 100
 * <p>
 * <p>
 * Related Topics Array Two Pointers 👍 21296 👎 5110
 */

package leetcode.editor.en;

import java.util.Arrays;

public class NextPermutation {
    public static void main(String[] args) {
        Solution solution = new NextPermutation().new Solution();
        int[] nums = new int[]{1, 3, 2};
        solution.nextPermutation(nums);
        System.out.println(Arrays.toString(nums));
        nums = new int[]{1, 1, 5};
        solution.nextPermutation(nums);
        System.out.println(Arrays.toString(nums));
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public void nextPermutation(int[] nums) {
            int i = nums.length - 2;

            for (; i >= 0 && nums[i] >= nums[i + 1]; i--) ;
            if (i >= 0) {
                //使i之后值最大
                int j = nums.length - 1;
                for (; j > 0 && nums[j] <= nums[i]; j--) ;
                swap(nums, i, j);
            }
            //使i之后值最小
            reverse(nums, i + 1, nums.length - 1);
        }

        private void swap(int[] arr, int i, int j) {
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }

        private void reverse(int[] nums, int l, int r) {
            while (l < r) {
                swap(nums, l++, r--);
            }
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}